sstables/mx/writer.cc

```diff
@@ -571,11 +571,10 @@ class writer : public sstable_writer::writer_impl {
     struct {
         // Unfortunately we cannot output the promoted index directly to the
         // index file because it needs to be prepended by its size.
-        // first_entry is used for deferring serialization into blocks for small partitions.
-        std::optional<pi_block> first_entry;
-        bytes_ostream blocks; // Serialized pi_blocks.
-        bytes_ostream offsets; // Serialized block offsets (uint32_t) relative to the start of "blocks".
-        uint64_t promoted_index_size = 0; // Number of pi_blocks inside blocks and first_entry;
+        // blocks is used for deferring serialization into blocks for small partitions.
+        utils::chunked_vector<pi_block> blocks;
```

Comment:
This is a regression in memory consumption for small promoted indexes since each pi_block involves two allocations, and allocations are rounded  up to a power of two. bytes_ostream switches to chunks of size 128k pretty soon.

I suggest to drop this patch. It means that the size if not constant, but it should be enough to keep it in check.

Reply:
Hmm, my first thought was to reparse the existing serialized blocks in order to down-sample them and reserialize them into a new bytes_ostream, but this is complicated (and cpu consuming).

IIUC, what you suggest is variable resolution, so we keep using the same block ostream but double the desired block size so the additional blocks will use half the resolution.  We can then increase the current (per-partition write) threshold so the next auto-scaling point will be when the block ostream will grow in another threshold bytes in size.

Reply:
Yes, access to the front of the partition will be faster than access into the back. But the motivation is to keep it simple and reduce chance of problems.